lib/enmeshed/connector.rb

```diff
@@ -0,0 +1,179 @@
+# frozen_string_literal: true
+
+module Enmeshed
+  class ConnectorError < StandardError; end
+
+  class Connector
+    # The app does not allow users to scan expired templates.
+    # However, previously scanned and then expired templates can still be submitted,
+    # resulting in the app silently doing nothing. CodeHarbor would still accept Relationships for expired templates if sent by the app.
+    # To minimize the risk of a template expiring before submission, we set the validity to 12 hours.
+    TEMPLATE_VALIDITY_PERIOD = 12.hours
+    # These attributes are mandatory in the app and must be provided.
+    # See https://enmeshed.eu/integrate/attribute-values for more attributes.
+    REQUIRED_ATTRIBUTES = %w[GivenName Surname EMailAddress AffiliationRole].freeze
+
+    API_KEY = Settings.omniauth&.nbp&.enmeshed&.connector_api_key
+    CONNECTOR_URL = Settings.omniauth&.nbp&.enmeshed&.connector_url
+    DISPLAY_NAME = Settings.omniauth&.nbp&.enmeshed&.display_name
+
+    @conn = nil
+    @enmeshed_address = nil
+    @display_name_id = nil
+
+    def self.create_relationship_template(nbp_uid)
+      new_template = parse_result(conn.post('/api/v2/RelationshipTemplates/Own') do |req|
+        req.body = relationship_template(nbp_uid).to_json
+      end)
+
+      Rails.logger.debug { "Enmeshed::ConnectorApi RelationshipTemplate created: #{new_template[:truncatedReference]}" }
+      RelationshipTemplate.new(new_template[:truncatedReference])
+    end
+
+    def self.pending_relationship_for_nbp_uid(nbp_uid)
+      relationships = parse_relationships(conn.get('/api/v2/Relationships') do |req|
```

Comment:
I just noticed that this route only returns useful results when the connector is synchronized (what a surprise 🙈). Synchronizing can be done with two mechanism:

1. Regular polling [configured through the `sync` module at the connector](https://enmeshed.eu/operate/configuration#sync).
2. Triggering a manual sync through [`/api/v2/Account/sync`](https://doc-demo.is.enmeshed.eu/docs/swagger/#/Account/sync).

As far as I can tell, both options can be combined together, but at least one is required for the system to work as expected.

My question is: Which mechanism should we use in our production setup? In development, we used the `sync` module with an interval of `10` seconds. So far, our production setup had the `sync` module disabled, following the enmeshed default config. (Hence, it didn't work as expected, since my `Relationship` got never synched). I see the following options:

1. Only rely on the `sync` module with a reasonable interval (anything up to `10` seconds, I'd say). It has a constant server load for enmeshed, and users are expected to wait a maximum of the `sync` `interval` (max. 10 seconds) + the frontend polling (1 second) = 11 seconds.
2. Only rely on the manual sync request through the API. This might be triggered before querying for `Relationships` (and might contain the new changes [Relationships, Messages] already, maybe saving a follow-up request to `/api/v2/Relationships`). Users can get a response faster.
3. Combine both mechanism together, regular synching through the module and manually through the API. This might still have the constant load on the enmeshed server, but also speed it for our users. Potentially, it would allow us to select a longer `interval`. While it might be more "fault tolerant" for our code, it also increases overall complexity.

Reply:
I would keep it simple and go with the sync module. I think a 5 seconds polling interval should be fine. We could also go lower though. I am honestly not too worried about the enmeshed server load because they can implement a polling-free sync module for their connectors anytime, right?

Reply:
Okay, I configured a polling interval of five seconds on our production systems.

Yes, the team behind enmeshed could refactor their sync module at any time and thereby reduce the server load "themselves". We have renovatebot enabled for the docker-compose.yml file, so that we'll get PR for new versions of the connector automatically. Usually, we aim to install these new versions timely.